packages/convex-helpers/server/filterWith.ts

```diff
@@ -0,0 +1,216 @@
+    }) as this;
+  }
+  order(order: "asc" | "desc"): QueryWithFilter<T> {
+    return new QueryWithFilter(this.q.order(order), this.p);
+  }
+  async paginate(
+    paginationOpts: PaginationOptions
+  ): Promise<PaginationResult<DocumentByInfo<T>>> {
+    const result = await this.q.paginate(paginationOpts);
+    return {...result, page: await asyncFilter(result.page, this.p)};
+  }
+  async collect(): Promise<DocumentByInfo<T>[]> {
+    const results = await this.q.collect();
+    return await asyncFilter(results, this.p);
+  }
+  async take(n: number): Promise<DocumentByInfo<T>[]> {
+    const results: DocumentByInfo<T>[] = [];
+    for await (const result of this) {
+      results.push(result);
+      if (results.length >= n) {
+        break;
+      }
+    }
+    return results;
+  }
+  async first(): Promise<DocumentByInfo<T> | null> {
+    for await (const result of this) {
+      return result;
+    }
+    return null;
+  }
+  async unique(): Promise<DocumentByInfo<T> | null> {
+    let uniqueResult: DocumentByInfo<T> | null = null;
+    for await (const result of this) {
+      if (uniqueResult === null) {
+        uniqueResult = result;
+      } else {
+        throw new Error("not unique");
+      }
+    }
+    return uniqueResult;
+  }
+  [Symbol.asyncIterator](): AsyncIterator<DocumentByInfo<T>, any, undefined> {
+    this.iterator = this.q[Symbol.asyncIterator]();
+    return this;
+  }
+  async next(): Promise<IteratorResult<any>> {
+    for (;;) {
+      const { value, done } = await this.iterator!.next();
+      if (value && (await this.p(value))) {
+        return { value, done };
+      }
+      if (done) {
+        return { value: null, done: true };
+      }
+    }
+  }
+  return() {
+    return this.iterator!.return!();
+  }
+}
+
+export class QueryInitializerWithFilter<T extends GenericTableInfo>
+  implements QueryInitializer<T>
+{
+  q: QueryInitializer<T>;
+  p: Predicate<T>;
+
+  constructor(q: QueryInitializer<T>, p: Predicate<T> | null = null) {
+    this.q = q;
+    this.p = p ?? (async (_) => true);
+  }
+  fullTableScan(): QueryWithFilter<T> {
+    return new QueryWithFilter(this.q.fullTableScan(), this.p);
+  }
+  withIndex<IndexName extends keyof Indexes<T>>(
+    indexName: IndexName,
+    indexRange?:
+      | ((
+          q: IndexRangeBuilder<DocumentByInfo<T>, NamedIndex<T, IndexName>, 0>
+        ) => IndexRange)
+      | undefined
+  ): Query<T> {
+    return new QueryWithFilter(this.q.withIndex(indexName, indexRange), this.p);
+  }
+  withSearchIndex<IndexName extends keyof SearchIndexes<T>>(
+    indexName: IndexName,
+    searchFilter: (
+      q: SearchFilterBuilder<DocumentByInfo<T>, NamedSearchIndex<T, IndexName>>
+    ) => SearchFilter
+  ): OrderedQuery<T> {
+    return new QueryWithFilter(
+      this.q.withSearchIndex(indexName, searchFilter),
+      this.p
+    );
+  }
+  filter(predicate: (q: FilterBuilder<T>) => Expression<boolean>): any {
```

Comment:
this unfortunately results in the rest of the chain being `any`. unfortunately it also can't be `this` since `QueryWithFilter` doesn't extend it

Reply:
can we make it be `this` and do a cast in the function?